src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseAsyncInAsync.cs

```diff
@@ -0,0 +1,265 @@
+            if (containingAnonymousFunction is not null)
+            {
+                return containingAnonymousFunction;
+            }
+
+            ISymbol containingSymbol = context.ContainingSymbol;
+            while (containingSymbol is not null && containingSymbol is not IMethodSymbol)
+            {
+                containingSymbol = containingSymbol.ContainingSymbol;
+            }
+
+            IMethodSymbol parentMethod = (IMethodSymbol)containingSymbol;
+            return parentMethod;
+        }
+
+        private static bool IsInTaskReturningMethodOrDelegate(OperationAnalysisContext context)
+        {
+            // We want to scan invocations that occur inside Task and Task<T>-returning delegates or methods.
+            // That is: methods that either are or could be made async.
+            IMethodSymbol parentMethod = GetParentMethodOrDelegate(context);
+            if (parentMethod == null)
+            {
+                return false;
+            }
+
+            ITypeSymbol returnTypeSymbol = parentMethod.ReturnType;
+
+            if (returnTypeSymbol is null)
+            {
+                return false;
+            }
+
+            return IsAsyncCompatibleReturnType(returnTypeSymbol);
+        }
+
+        private static readonly IReadOnlyList<string> SystemRuntimeCompilerServices = new[]
+        {
+            nameof(System),
+            nameof(System.Runtime),
+            nameof(System.Runtime.CompilerServices),
+        };
+
+        private static bool IsAsyncCompatibleReturnType(ITypeSymbol? typeSymbol)
+        {
+            if (typeSymbol is null)
+            {
+                return false;
+            }
+
+            // ValueTask and ValueTask<T> have the AsyncMethodBuilderAttribute
+            return (typeSymbol.Name == nameof(Task) && typeSymbol.BelongsToNamespace(Namespaces.SystemThreadingTasks))
+                || IsIAsyncEnumerable(typeSymbol) || typeSymbol.AllInterfaces.Any(IsIAsyncEnumerable)
+                || typeSymbol.GetAttributes().Any(ad => ad.AttributeClass?.Name == nameof(System.Runtime.CompilerServices.AsyncMethodBuilderAttribute) &&
+                ad.AttributeClass.BelongsToNamespace(SystemRuntimeCompilerServices));
+
+            static bool IsIAsyncEnumerable(ITypeSymbol symbol)
+                => symbol.Name == "IAsyncEnumerable"
+                && symbol.BelongsToNamespace(Namespaces.SystemCollectionsGeneric);
+        }
+
+        private static bool InspectMemberAccess(OperationAnalysisContext context, IEnumerable<CommonInterests.SyncBlockingMethod> problematicMethods)
+        {
+            ISymbol? memberSymbol = context.Operation.SemanticModel.GetSymbolInfo(context.Operation.Syntax, context.CancellationToken).Symbol;
+            if (memberSymbol is object)
+            {
+                foreach (CommonInterests.SyncBlockingMethod item in problematicMethods)
+                {
+                    if (item.Method.IsMatch(memberSymbol))
```

Comment:
So this is going through a list/array each time to match a method. It might be faster to use a HashSet here? But the list only has 5 members in it, so I wouldn't bother about it too much. Just something to keep in mind if it comes up in the review. 

If you wanted to change it to a HashSet, I think you instantiate it in the normal way and pass in ISymbolComparison.Default(I think that's the name) as the comparator